google/cloud/pubsub/internal/tracing_message_callback_test.cc

```diff
@@ -0,0 +1,105 @@
+#include "google/cloud/internal/opentelemetry.h"
+#include "google/cloud/testing_util/is_proto_equal.h"
+#include "google/cloud/testing_util/opentelemetry_matchers.h"
+#include "google/cloud/testing_util/status_matchers.h"
+#include <opentelemetry/context/propagation/text_map_propagator.h>
+#include <opentelemetry/trace/propagation/http_trace_context.h>
+#include <opentelemetry/trace/scope.h>
+#include <opentelemetry/trace/semantic_conventions.h>
+#endif  // GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+#include <gmock/gmock.h>
+
+namespace google {
+namespace cloud {
+namespace pubsub_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+using ::google::cloud::testing_util::InstallSpanCatcher;
+using ::google::cloud::testing_util::OTelAttribute;
+using ::google::cloud::testing_util::SpanHasAttributes;
+using ::google::cloud::testing_util::SpanHasInstrumentationScope;
+using ::google::cloud::testing_util::SpanKindIsInternal;
+using ::google::cloud::testing_util::SpanNamed;
+using ::google::cloud::testing_util::SpanWithParent;
+using ::testing::AllOf;
+using ::testing::Contains;
+
+pubsub::Subscription TestSubscription() {
+  return pubsub::Subscription("test-project", "test-sub");
+}
+
+std::shared_ptr<MessageCallback> MakeTestMessageCallback(
+    std::shared_ptr<MessageCallback> mock) {
+  return MakeTracingMessageCallback(
+      std::move(mock),
+      Options{}.set<pubsub::SubscriptionOption>(TestSubscription()));
+}
+
+TEST(TracingMessageCallback, UserCallback) {
+  namespace sc = opentelemetry::trace::SemanticConventions;
+  auto span_catcher = InstallSpanCatcher();
+  auto mock = std::make_shared<pubsub_testing::MockMessageCallback>();
+  EXPECT_CALL(*mock, user_callback).Times(1);
+  auto message_callback = MakeTestMessageCallback(std::move(mock));
+
+  auto span = internal::MakeSpan("my-sub subscribe");
+  MessageCallback::MessageAndHandler m{
+      pubsub::MessageBuilder().Build(),
+      std::make_unique<pubsub_testing::MockExactlyOnceAckHandlerImpl>(),
+      "ack-id",
+      {span}};
+  message_callback->user_callback(std::move(m));
+  span->End();
+
+  auto spans = span_catcher->GetSpans();
+  EXPECT_THAT(
+      spans, Contains(AllOf(SpanHasInstrumentationScope(), SpanKindIsInternal(),
+                            SpanNamed("test-sub process"),
+                            SpanHasAttributes(OTelAttribute<std::string>(
+                                sc::kMessagingSystem, "gcp_pubsub")),
+                            SpanWithParent(span))));
+}
+
+#else  // GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+TEST(TracingMessageCallback,
+     VerifyMessageCallbackIsNotNullWhenOTelIsNotCompiled) {
+  auto message_callback = MakeTracingMessageCallback(
+      std::make_shared<pubsub_testing::MockMessageCallback>(),
+      Options{}.set<pubsub::SubscriptionOption>(
+          pubsub::Subscription("test-project", "test-sub")));
+
+  EXPECT_THAT(message_callback, testing::Not(testing::IsNull()));
```

Comment:
same nit about `using ::testing::Foo;`

but..... this test could be stricter. I know I said "make sure we are not returning `nullptr`". I should have said "Make sure we are using the supplied `message_callback`".

I think you should call `message_callback->user_callback()` and verify that the mock gets called.

Reply:
Done